### PR TITLE
Adds vagrant runner

### PR DESCRIPTION
**Pull request by JayjeetAtGithub**
Addresses #172

### CODE REVIEW

**Comment by JayjeetAtGithub** [comment]
@ivotron While adding `VagrantRunner`,  in implementing the `--reuse` feature, we need to check the existence of a VM. For this, we need a vagrant cli functionality `vagrant global-status` which is absent in `python-vagrant`. I forked `python-vagrant` and implemented the feature and then used it to build and test `VagrantRunner` temporarily. Also, this is due to which the builds are failing. So, what shall we do now, continue with the forked lib or is there any other way around? Please have a look. Thanks.

**Comment by JayjeetAtGithub** [comment]
@ivotron `virtualbox` installation is failing in Travis CI even after several different attempts. Some answers say that we can't run `virtualbox` in Travis CI. Can you please help me out with this?

**Comment by ivotron** [comment]
hey @JayjeetAtGithub, thanks a lot for working on this. We have also tried to install virtualbox on Travis in the past without success. If this gets too much of your time, then I'd say we leave it out for now.

Regarding the changes done to the `python-vagrant` package, we can open a PR on the upstream repo. We can point to the fork in the `setup.py` file in the meantime (hopefully they'll merge the PR).

**Comment by JayjeetAtGithub** [comment]
@ivotron Does that mean we won't have a `test vagrant` in the codebase and `VagrantRunner` would depend on local machine tests?

**Comment by ivotron** [comment]
@JayjeetAtGithub , yes, that would mean we have to test manually. Alternatively, we could also see if we can use the docker provider for vagrant. In this way we test the popper/vagrant interaction.

**Comment by JayjeetAtGithub** [comment]
@ivotron Can we clone the remote actions within the runners ?. This is needed to reuse the environment in case of vagrant which otherwise gets destroyed during erasing the prev clone and recloning. We would like to check whether the prev clone is present and if present we would skip cloning.

**Comment by JayjeetAtGithub** [comment]
@ivotron Can you also give me some idea how to set the env vars inside the vagrant VM's ?
I was doing `vagrant ssh --command "echo export VAR=value >> ~/.bashrc"`, but this was'nt working.

**Comment by ivotron** [comment]
> @ivotron Can you also give me some idea how to set the env vars inside the vagrant VM's ?
> I was doing `vagrant ssh --command "echo export VAR=value >> ~/.bashrc"`, but this was'nt working.

I believe this can be done by modifying the `Vagrantfile`, so that we add a shell provisioner and define the environment variables there.

**Comment by ivotron** [comment]
> @ivotron Can we clone the remote actions within the runners ? This is needed to reuse the environment in case of vagrant which otherwise gets destroyed during erasing the prev clone and recloning. We would like to check whether the prev clone is present and if present we would skip cloning when the `--reuse` flag is passed. In this way, we can preserve the vagrant environment for reusing.

This is related to #550 and #652, in the sense that if those two were already solved, the problem you described would not arise.

**Comment by JayjeetAtGithub** [comment]
@ivotron Please review this. The proper functioning of this feature depends on #653. Thanks.

File: cli/popper/gha.py [hunk 1]
```diff
@@ -537,10 +559,114 @@ def singularity_start(self):
...
+            log.fail('Action {} failed!'.format(self.action['name']))
+
+        cmd = []
+        if self.action.get('runs', None):
+            cmd.extend(self.action['runs'])
+        if self.action.get('args', None):
+            cmd.extend(self.action['args'])
+        self.vagrant_ssh(' '.join(cmd))
+
+    def setenv(self, k, v):
```
Comment (ivotron) [inline cli/popper/gha.py]: looks like this is not used, right? if it isn't, can we remove it?

Comment (JayjeetAtGithub) [reply cli/popper/gha.py]: > Also, we need to deal with sharing $HOME and $GITHUB_WORKSPACE folders.

Can you please help me on how to pass these env vars inside the vagrant vm automatically? I was thinking of making a template Vagrantfile and populating it during `vagrant_init()`. Is there any other way round ?

Comment (ivotron) [reply cli/popper/gha.py]: I do not think there is a workaround it. We will need to dynamically inject this into the Vagrantfile.

Comment (JayjeetAtGithub) [reply cli/popper/gha.py]: Yeah, thanks

**Pull request closed without merging**